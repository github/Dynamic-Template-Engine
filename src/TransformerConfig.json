{
  "cardRenderer":[
    {
        "SourceType": "IssueOpened",
        "ClientType": "Teams",
        "TemplateType": "HandleBars",
        "TemplateName": "issue_opened.handlebars"
    },
    {
        "SourceType": "PullRequestOpened",
        "ClientType": "Teams",
        "TemplateType": "HandleBars",
        "TemplateName": "issue_opened.handlebars"
    },
    {
        "SourceType": "IssueOpened",
        "ClientType": "Teams",
        "TemplateType": "Liquid",
        "TemplateName": "issue_opened.liquid"
    }
  ],
  "partials":[],
  "eventTransformer":[
    {
      "SourceType":"IssueOpened",
      "TemplateType":"HandleBars",
      "TemplateName":"issue_opened.handlebars"
    }
  ]
}