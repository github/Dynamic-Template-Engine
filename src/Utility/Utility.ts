/** Copyright (c) 2020 GitHub. This code is licensed under MIT license (see LICENSE(https://github.com/github/event-transformer/blob/feature/chatops/LICENSE) for details) */
import * as https from 'https';
import * as fs from 'fs';
import * as path from 'path';
import { Octokit } from '@octokit/rest';
import { EmptyFileError, FileReadError } from '../Error/FileError';

/**
 * Utility functions available to the whole code base
 */
export default class Utility {
  /**
   * A synchoronous http call
   *
   * @param {string} url - url to make the http call to
   * @returns {string} response string
   */
  public static httpSync(url: string) : Promise<string> {
    return new Promise<string>((resolve, reject) => {
      https.get(url, (response) => {
        const { statusCode } = response;
        if (statusCode !== 200) {
          reject(new Error(`Http Call failed. Status Code: ${statusCode}`));
        }
        const chunks_of_data: Buffer[] = [];

        response.on('data', (fragments) => {
          chunks_of_data.push(fragments);
        });

        response.on('end', () => {
          const response_body = Buffer.concat(chunks_of_data);
          resolve(response_body.toString());
        });

        response.on('error', (error) => {
          reject(error);
        });
      });
    });
  }

  /**
   * Fetch file either from local machine or using an http call
   *
   * @param {boolean} fromRepo - is an from repo or a local machine lookup
   * @param {string} repo - name of the repository
   * @param {boolean} branch - name of the branch
   * @param {string} filePath - the path of the file to read
   */
  public static async fetchFile(fromRepo: boolean, repo: string, branch: string, filePath: string,
    accessToken?: string): Promise<string> {
    let file = '';
    try {
      if (fromRepo) {
        file = await this.getFile(repo, branch, filePath, accessToken);
      } else {
        file = fs.readFileSync(path.resolve(__dirname, `../${filePath}`)).toString();
      }
    } catch (error) {
      throw new FileReadError(`Could not read file with 
        file path: ${filePath} \n ERROR: ${error.message}`);
    }
    if (!file || file.length <= 0) {
      throw new EmptyFileError(`Empty file at path: ${filePath}`);
    }
    return file;
  }

  /**
   * Creates a key based on all strings passed
   * Implementation can be changed to create a hash of all strings rather than just concating
   *
   * @param {string[]} allStrings - all strings to create a unique key
   * @returns {string} generated key
   */
  public static keyGenerator(...allStrings: string[]): string {
    let genratedKey = '';
    allStrings.forEach(element => {
      genratedKey += `${element}.`;
    });
    return genratedKey.slice(0, genratedKey.length - 1);
  }

  public static async getFile(repo: string, branch: string, filePath: string, token?: string) {
    try {
      const client = new Octokit({
        auth: token,
      });
      const ownerName = repo.split('/')[0];
      const repoName = repo.split('/')[1];
      const response = await client.repos.getContent({
        owner: ownerName,
        repo: repoName,
        path: filePath,
        ref: branch,
      });
      // eslint-disable-next-line prefer-destructuring
      const data : any = response.data;
      if (!data.content || !data.encoding) {
        throw new Error('Could not fetch file contents');
      }
      const template = Buffer.from(data.content, data.encoding as BufferEncoding).toString();
      return template;
    } catch (error) {
      throw new Error(error.message);
    }
  }
}
