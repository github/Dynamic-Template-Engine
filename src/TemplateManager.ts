// Copyright (c) 2020 GitHub. This code is licensed under MIT license (see LICENSE(https://github.com/github/event-transformer/blob/feature/chatops/LICENSE) for details)
/* eslint-disable @typescript-eslint/no-throw-literal */
import * as path from 'path';
import Transformer from './Transformer/Core/Transformer';
import TransformerConfig from './Transformer/Model/TransformerConfig';
import Utils from './Utility/Utility';
import BaseTransformConfigEntry from './Transformer/Model/BaseTransformConfigEntry';
import CardRenderer from './Transformer/CardRenderer/CardRenderer';
import EventTransformer from './Transformer/EventTransformer/EventTransformer';
import { TemplateParseError, TemplateEngineNotFound } from './Error/TemplateError';
import { FileParseError, EmptyFileError } from './Error/FileError';

/**
 * Template Manager provides methods to setup the template configuration
 * intializes template engines, registers all the templates provided by the config
 */
export default class TemplateManager {
  /**
   * Sets up the templates, picking from the path provided,
   * registering them to appropriate engines
   *
   * @param {string} configFilePath - path of the config file
   * @returns {boolean} true if setup successful
   * @throws Error if setup fails
   */
  public static async setupTemplateConfiguration(configFilePath: string): Promise<boolean> {
    try {
      const baseUrl = configFilePath.substring(0, configFilePath.lastIndexOf(path.sep));
      const transformerConfig = await this.readConfigFile(configFilePath, false);
      await this.registerAllTemplates(baseUrl, new CardRenderer(), transformerConfig.cardRenderer);
      await this.registerAllTemplates(baseUrl, new EventTransformer(),
        transformerConfig.eventTransformer);
    } catch (error) {
      if (error instanceof TemplateEngineNotFound || error instanceof TemplateParseError
        || error instanceof FileParseError || error instanceof EmptyFileError) {
        throw error;
      } else {
        throw new Error(`Setup failed for the config file path: ${configFilePath} \n ERROR: ${error.message}`);
      }
    }
    return true;
  }

  /**
   * Sets up the templates, picking from the path provided,
   * registering them to appropriate engines
   *
   * @param {string} repo - repo name ex. user/repo
   * @param {string} branch - ex. master
   * @param {string} configName config file name in the repo root folder
   * @returns {boolean} true if setup succesful
   * @throws Error if setup fails
   */
  public static async setupTemplateConfigurationFromRepo(repo: string, branch: string,
    configName: string): Promise<boolean> {
    const baseUrl = `https://raw.githubusercontent.com/${repo}/${branch}`;
    try {
      const transformerConfig = await this.readConfigFile(`${baseUrl}/${configName}.json`, true);
      await this.registerAllTemplates(baseUrl, new CardRenderer(),
        transformerConfig.cardRenderer);
      await this.registerAllTemplates(baseUrl, new EventTransformer(),
        transformerConfig.eventTransformer);
    } catch (error) {
      if (error instanceof TemplateEngineNotFound || error instanceof TemplateParseError
        || error instanceof FileParseError || error instanceof EmptyFileError) {
        throw error;
      } else {
        throw new Error(`Setup failed. \n ERROR: ${error.message}`);
      }
    }
    return true;
  }

  /**
   * Read config file and deserialize the file appropriately
   *
   * @param {string} filePath - file path of the config
   * @param {boolean} fromRepo - specifies if file from repo or from local machine
   */
  private static async readConfigFile(filePath: string,
    fromRepo: boolean): Promise<TransformerConfig> {
    const data = await Utils.fetchFile(fromRepo, filePath);
    try {
      return <TransformerConfig>JSON.parse(data.toString());
    } catch (error) {
      throw new FileParseError(`Unable to parse config file from path 
      ${filePath}, original error message: ${error.message}`);
    }
  }

  /**
   * Register all templates provided in the transformerConfig
   *
   * @param {string} baseUrl - base url for the template files
   * @param {string} transformer - transformer whith which template should be registered
   * @param {BaseTransformConfigEntry} transformerConfigs - the template transformer configs
   */
  private static async registerAllTemplates(baseUrl: string, transformer: Transformer<any>,
    transformerConfigs: BaseTransformConfigEntry[]): Promise<void> {
    // eslint-disable-next-line no-restricted-syntax
    for (const element of transformerConfigs) {
      try {
        // eslint-disable-next-line no-await-in-loop
        await transformer.registerTemplate(baseUrl, element);
      } catch (error) {
        if (error instanceof TemplateParseError) {
          throw new TemplateParseError(`Failed to parse template with name: ${element.TemplateName} 
          for source type: ${element.SourceType} and template type: ${element.TemplateType}`);
        } else {
          throw error;
        }
      }
    }
  }
}
