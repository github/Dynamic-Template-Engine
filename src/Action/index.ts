import * as core from '@actions/core';
import * as github from '@actions/github';
import CardRenderer from '../Transformer/CardRenderer/CardRenderer';
import EventTransformer from '../Transformer/EventTransformer/EventTransformer';
import TemplateManager from '../TemplateManager';
import { ClientType, TemplateType } from '../Transformer/Core/TransformContract';

const TemplateTypeMap = new Map<string, TemplateType>(Object.entries(TemplateType).map(
  ([key, value]:[string, TemplateType]) => [key, value],
));

const ClientTypeMap = new Map<string, ClientType>(Object.entries(ClientType).map(
  ([key, value]:[string, ClientType]) => [key, value],
));

function throwIfUndefined<T>(value: T|undefined): T {
  if (value) {
    return value;
  }
  throw new Error('Undefined value found');
}

async function run(): Promise<void> {
  try {
    const options: core.InputOptions = { required: true };
    const repoName: string = core.getInput('repoName', options);
    const branch: string = core.getInput('branchName', options);
    const templateTypeString = core.getInput('templateType', options);
    const sourceType: string = core.getInput('sourceType', options);
    const clientTypeString: string = core.getInput('clientType');
    const accessToken: string = core.getInput('accessToken');
    const data: string = JSON.stringify(github.context.payload, undefined, 2);
    const dataJson: JSON = JSON.parse(data);
    const templateType: TemplateType = throwIfUndefined<TemplateType>(
      TemplateTypeMap.get(templateTypeString),
    );
    let clientType: ClientType | undefined;
    if (clientTypeString) {
      clientType = throwIfUndefined<ClientType>(
        ClientTypeMap.get(clientTypeString),
      );
    }
    let renderedTemplate: string;
    await TemplateManager.setupTemplateConfigurationFromRepo(repoName, branch, sourceType,
      templateType, clientType, accessToken);
    if (clientType != null) {
      const cardRenderer = new CardRenderer();
      renderedTemplate = await cardRenderer.ConstructCardJson(templateType, sourceType, clientType,
        dataJson);
    } else {
      const eventTransformer = new EventTransformer();
      renderedTemplate = await eventTransformer.ConstructEventJson(templateType, sourceType,
        dataJson);
    }
    renderedTemplate = JSON.parse(renderedTemplate);
    core.setOutput('renderedTemplate', renderedTemplate);
    const octokit = github.getOctokit(accessToken);
    const { owner, repo } = github.context.repo;
    const event_type = 'custom';
    octokit.repos.createDispatchEvent({
      owner,
      repo,
      event_type,
      client_payload: renderedTemplate,
    });
  } catch (error) {
    core.setFailed(error);
  }
}

run();
