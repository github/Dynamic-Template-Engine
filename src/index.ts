// Copyright (c) 2020 GitHub. This code is licensed under MIT license (see LICENSE(https://github.com/github/event-transformer/blob/feature/chatops/LICENSE) for details)
import TemplateManager from './TemplateManager';
import CardRenderer from './Transformer/CardRenderer/CardRenderer';
import EventTransformer from './Transformer/EventTransformer/EventTransformer';

export { TemplateManager };
export { CardRenderer, EventTransformer };
export { ClientType, TemplateType } from './Transformer/Core/TransformContract';
