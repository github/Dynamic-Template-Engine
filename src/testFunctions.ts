import { TemplateManager } from './TemplateManager';
import { CardRenderer } from './Transformer/CardRenderer/CardRenderer';
import { TemplateType, ClientType } from './Transformer/Core/TransformContracts';
import { testData } from './handlebars-test';

export async function test(){
  const path = 'TransformerConfig.json';
  //await TemplateManager.setupTemplateConfiguration(path);
  await TemplateManager.setupTemplateConfigurationFromRepo('anuraag016/TestTemplateEngine', 'master', 'TransformerConfig');
  testingTemplateRender();
  //setTimeout(testingTemplateRender, 5000);
}

const testingTemplateRender = ()=>{
  const cardRenderer = new CardRenderer();
  const renderedCard = cardRenderer.ConstructCardJson(TemplateType.HandleBars, "IssueOpened", ClientType.Teams, testData);
  console.log("START : Handlebars render card::::::::::::::::::::::::::::")
  console.log(renderedCard);
  console.log("END : Handlebars render card::::::::::::::::::::::::::::")
  const renderedCard2 = cardRenderer.ConstructCardJson(TemplateType.Liquid, "IssueOpened", ClientType.Teams, testData);
  console.log("START : Liquid render card::::::::::::::::::::::::::::")
  console.log(renderedCard2);
  console.log("END : Liquid render card::::::::::::::::::::::::::::")
};