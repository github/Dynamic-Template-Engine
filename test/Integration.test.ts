import { TemplateManager, CardRenderer, EventTransformer, TemplateType, ClientType } from "../src/index";
import * as nock from "nock";
import * as MockTransformerConfig from './MockTemplate/MockTransformerConfig.json';
import Utility from "../src/Utility/Utility";
import * as path from 'path';

describe('End to end test with mock data and mock templates from local filesystem', () => {

  const mockData = {
    post: {
      name: "Test Post",
      description: "This is mock data of a post",
      link: "http://ItsMockData.com",
      createDate: "1594060200000"
    },
    owner: {
      name: "Tester",
      handle: "tester005",
    }
  }

  beforeAll(async () => {
    await TemplateManager.setupTemplateConfiguration(path.resolve(__dirname, 'MockTemplate/MockTransformerConfig.json'));
  });

  it.each`
  templateType | clientType | sourceType
  ${TemplateType.HandleBars} | ${ClientType.Teams} | ${"MockData"}
  ${TemplateType.Liquid} | ${ClientType.Teams} | ${"MockData"}
  `('Test CardRenderer for $templateType and $clientType', ({ templateType, clientType, sourceType }) => {
    const cardRenderer = new CardRenderer();
    expect(cardRenderer.ConstructCardJson(templateType, sourceType, clientType, { 'name': 'Test' })).toMatchSnapshot();
  });

  it.each`
  templateType | sourceType
  ${TemplateType.HandleBars} | ${"MockData2"}
  ${TemplateType.Liquid} | ${"MockData2"}
  `('Test EventTransformer for $templateType', ({ templateType, sourceType }) => {
    const eventTransformer = new EventTransformer();
    expect(eventTransformer.ConstructEventJson(templateType, sourceType, { 'name': 'Test' })).toMatchSnapshot();
  });

  it.each`
  templateType | clientType | sourceType
  ${TemplateType.HandleBars} | ${ClientType.Teams} | ${"MockPostData"}
  ${TemplateType.Liquid} | ${ClientType.Teams} | ${"MockPostData"}
  `('Test CardRenderer for $templateType and $clientType', ({ templateType, clientType, sourceType }) => {
    const cardRenderer = new CardRenderer();
    expect(cardRenderer.ConstructCardJson(templateType, sourceType, clientType, mockData)).toMatchSnapshot();
  });
});

describe('End to end test with mock data and mock templates from github repos', () => {

  const mockData = {
    post: {
      name: "Test Post",
      description: "This is mock data of a post",
      link: "http://ItsMockData.com",
      createDate: "1594060200000"
    },
    owner: {
      name: "Tester",
      handle: "tester005",
    }
  }

  const mockRepoName = 'test/mockTemplates';
  const mockBranchName = 'master';
  const mockTransformerConfigName = 'MockTransformerConfig';

  const setupMockApiCalls = () => {
    MockTransformerConfig.cardRenderer.forEach(element => {
      nock("https://raw.githubusercontent.com")
        .get(`/${mockRepoName}/${mockBranchName}/CardTemplate/${element.ClientType}/${element.TemplateType}/${element.TemplateName}`)
        .reply(200,
          async () => {
            return await Utility.fetchFile(false, path.resolve(__dirname, `MockTemplate/CardTemplate/${element.ClientType}/${element.TemplateType}/${element.TemplateName}`));
          }
        );
    });
    MockTransformerConfig.eventTransformer.forEach(element => {
      nock("https://raw.githubusercontent.com")
        .get(`/${mockRepoName}/${mockBranchName}/EventTemplate/${element.TemplateType}/${element.TemplateName}`)
        .reply(200, 
          async () => { 
            return await Utility.fetchFile(false, path.resolve(__dirname, `MockTemplate/EventTemplate/${element.TemplateType}/${element.TemplateName}`));
          }
        );
    });
  };

  beforeAll(async () => {
    nock("https://raw.githubusercontent.com")
      .get(`/${mockRepoName}/${mockBranchName}/${mockTransformerConfigName}.json`).reply(200, JSON.stringify(MockTransformerConfig));

    setupMockApiCalls();
    await TemplateManager.setupTemplateConfigurationFromRepo(mockRepoName, mockBranchName, mockTransformerConfigName);
  });

  it.each`
  templateType | clientType | sourceType
  ${TemplateType.HandleBars} | ${ClientType.Teams} | ${"MockData"}
  ${TemplateType.Liquid} | ${ClientType.Teams} | ${"MockData"}
  `('Test CardRenderer for $templateType and $clientType', ({ templateType, clientType, sourceType }) => {
    const cardRenderer = new CardRenderer();
    expect(cardRenderer.ConstructCardJson(templateType, sourceType, clientType, { 'name': 'Test' })).toMatchSnapshot();
  });

  it.each`
  templateType | sourceType
  ${TemplateType.HandleBars} | ${"MockData2"}
  ${TemplateType.Liquid} | ${"MockData2"}
  `('Test EventTransformer for $templateType', ({ templateType, sourceType }) => {
    const eventTransformer = new EventTransformer();
    expect(eventTransformer.ConstructEventJson(templateType, sourceType, { 'name': 'Test' })).toMatchSnapshot();
  });

  it.each`
  templateType | clientType | sourceType
  ${TemplateType.HandleBars} | ${ClientType.Teams} | ${"MockPostData"}
  ${TemplateType.Liquid} | ${ClientType.Teams} | ${"MockPostData"}
  `('Test CardRenderer for $templateType and $clientType', ({ templateType, clientType, sourceType }) => {
    const cardRenderer = new CardRenderer();
    expect(cardRenderer.ConstructCardJson(templateType, sourceType, clientType, mockData)).toMatchSnapshot();
  });
});