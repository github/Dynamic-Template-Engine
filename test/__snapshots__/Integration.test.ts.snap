// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`End to end test with mock data and mock templates from github repos Test CardRenderer for HandleBars and Teams 1`] = `"Hello Test"`;

exports[`End to end test with mock data and mock templates from github repos Test CardRenderer for HandleBars and Teams 2`] = `
"{
  \\"title\\": \\"Test Post\\",
  \\"author\\": \\"Tester\\"
  \\"createdOn\\": \\"1594060200000\\",
  \\"text\\": \\"This is mock data of a post\\" 
}"
`;

exports[`End to end test with mock data and mock templates from github repos Test CardRenderer for Liquid and Teams 1`] = `"Hello TEST"`;

exports[`End to end test with mock data and mock templates from github repos Test CardRenderer for Liquid and Teams 2`] = `
"{
  \\"title\\": \\"TEST POST\\",
  \\"author\\": \\"Tester\\"
  \\"createdOn\\": \\"1594060200000\\",
  \\"text\\": \\"This is mock data of a post\\" 
}"
`;

exports[`End to end test with mock data and mock templates from github repos Test EventTransformer for HandleBars 1`] = `"Hello Test"`;

exports[`End to end test with mock data and mock templates from github repos Test EventTransformer for Liquid 1`] = `"Hello test"`;

exports[`End to end test with mock data and mock templates from local filesystem Test CardRenderer for HandleBars and Teams 1`] = `"Hello Test"`;

exports[`End to end test with mock data and mock templates from local filesystem Test CardRenderer for HandleBars and Teams 2`] = `
"{
  \\"title\\": \\"Test Post\\",
  \\"author\\": \\"Tester\\"
  \\"createdOn\\": \\"1594060200000\\",
  \\"text\\": \\"This is mock data of a post\\" 
}"
`;

exports[`End to end test with mock data and mock templates from local filesystem Test CardRenderer for Liquid and Teams 1`] = `"Hello TEST"`;

exports[`End to end test with mock data and mock templates from local filesystem Test CardRenderer for Liquid and Teams 2`] = `
"{
  \\"title\\": \\"TEST POST\\",
  \\"author\\": \\"Tester\\"
  \\"createdOn\\": \\"1594060200000\\",
  \\"text\\": \\"This is mock data of a post\\" 
}"
`;

exports[`End to end test with mock data and mock templates from local filesystem Test EventTransformer for HandleBars 1`] = `"Hello Test"`;

exports[`End to end test with mock data and mock templates from local filesystem Test EventTransformer for Liquid 1`] = `"Hello test"`;
